src/neuroconv/tools/testing/data_interface_mixins.py

```diff
@@ -0,0 +1,163 @@
+import json
+from abc import abstractmethod
+from datetime import date, datetime
+from pathlib import Path
+from typing import Callable, Union
+
+import numpy as np
+from jsonschema import validate
+from jsonschema.validators import Draft7Validator
+from roiextractors import NwbImagingExtractor, NwbSegmentationExtractor
+from roiextractors.testing import check_imaging_equal, check_segmentations_equal
+from spikeinterface.core.testing import check_recordings_equal, check_sortings_equal
+from spikeinterface.extractors import NwbRecordingExtractor, NwbSortingExtractor
+
+from ...basedatainterface import BaseDataInterface
+from ...datainterfaces.ecephys.baserecordingextractorinterface import (
+    BaseRecordingExtractorInterface,
+)
+from ...datainterfaces.ecephys.basesortingextractorinterface import (
+    BaseSortingExtractorInterface,
+)
+from ...datainterfaces.ophys.baseimagingextractorinterface import (
+    BaseImagingExtractorInterface,
+)
+from ...datainterfaces.ophys.basesegmentationextractorinterface import (
+    BaseSegmentationExtractorInterface,
+)
+
+
+def json_serial(obj):
+    """JSON serializer for objects not serializable by default json code"""
+
+    if isinstance(obj, (datetime, date)):
+        return obj.isoformat()
+    raise TypeError("Type %s not serializable" % type(obj))
+
+
+class DataInterfaceTestMixin:
+    data_interface_cls: Union[BaseDataInterface, Callable]
+    interface_kwargs: Union[dict, list]
+    save_directory: Path
+
+    def test_source_schema_valid(self):
+        schema = self.data_interface_cls.get_source_schema()
+        Draft7Validator.check_schema(schema=schema)
+
+    def check_conversion_options_schema_valid(self):
+        schema = self.interface.get_conversion_options_schema()
+        Draft7Validator.check_schema(schema=schema)
+
+    def check_metadata_schema_valid(self):
+        schema = self.interface.get_metadata_schema()
+        Draft7Validator.check_schema(schema=schema)
+
+    def check_metadata(self):
+        schema = self.interface.get_metadata_schema()
+        metadata = self.interface.get_metadata()
+        if "session_start_time" not in metadata["NWBFile"]:
+            metadata["NWBFile"].update(session_start_time=datetime.now().astimezone())
+        # handle json encoding of datetimes and other tricky types
+        metadata_for_validation = json.loads(json.dumps(metadata, default=json_serial))
+        validate(metadata_for_validation, schema)
+
+        self.check_extracted_metadata(metadata)
+
+    def run_conversion(self, nwbfile_path: str):
+        metadata = self.interface.get_metadata()
+        metadata["NWBFile"].update(session_start_time=datetime.now().astimezone())
+        self.interface.run_conversion(nwbfile_path=nwbfile_path, overwrite=True, metadata=metadata)
+
+    @abstractmethod
+    def check_read(self, nwbfile_path: str):
+        pass
+
+    def check_extracted_metadata(self, metadata: dict):
+        pass
```

Comment:
OK, not required...

But this is what someone would override any time they want to make assertions on extracted metadata structure?

Reply:
Oh, actually the docstring for the method might be even better - that way the user can run `.check_extracted_metadata?` and it can instruct them what to do (and maybe show a small example of what such an assertion could include)